### PR TITLE
Implement first-light location refresh loop

### LINKED ISSUE #771: catalog backend: implement refresh loops
Implement the logic of the backend service that keeps components refreshed from descriptor files in a version control system.

## Background

The corresponding service at Spotify stores the latest known good version of component files in a data store, and continuously attempts to refresh them with new data from our GitHub Enterprise, via polling. It also keeps track of the current status of this refresh loop per component, along with errors encountered. This status is part of the response when fetching components.

The first version of the catalog service will do the same. Push based or event triggered updates will come later.

## Some Questions

- As a location is registered, shall an initial refresh attempt be made, and the location registration fail if the refresh fails?
- How will later failed refreshes be handled? What metadata about the errors shall we store and how are they communicated to callers or component owners?
- What version control systems shall we support? Presumably at the very least github and gitlab. Shall this be pluggable?
- How shall we handle authentication for requesting data from the version control system? Can we initially use `/raw/` or similar browser endpoints, or should we for example require that the user supply a dev token to pass to API calls?
- How shall we handle rate limits and other exceptional states when doing refreshes?

### PR DESCRIPTION
**Pull request by freben**
Fixes #771

This is the bare bones first iteration of things.

### CODE REVIEW

File: plugins/catalog-backend/src/catalog/DatabaseCatalog.ts [hunk 1]
```diff
@@ -15,29 +15,79 @@
...
     await database.migrate.latest({
       directory: path.resolve(__dirname, '..', 'migrations'),
       loadExtensions: ['.js'],
     });
 
-    return new DatabaseCatalog(database);
+    const databaseCatalog = new DatabaseCatalog(database, logger);
+
+    const startRefresh = async () => {
+      for (;;) {
```
Comment (nikek) [inline plugins/catalog-backend/src/catalog/DatabaseCatalog.ts:70]: `while (true)` might be less cryptic?

Comment (freben) [reply plugins/catalog-backend/src/catalog/DatabaseCatalog.ts:70]: That triggers an [eslint warning](https://eslint.org/docs/rules/no-constant-condition), and I would rather avoid the whole issue than excluding/ignoring the rule.

File: plugins/catalog-backend/src/catalog/DatabaseCatalog.ts [hunk 2]
```diff
@@ -15,29 +15,79 @@
...
+} from './types';
+import YAML from 'yaml';
+
+async function readLocation(location: Location): Promise<Component[]> {
+  switch (location.type) {
+    case 'file': {
+      let parsed;
+      try {
+        const raw = await fs.readFile(location.target, 'utf8');
+        parsed = YAML.parse(raw);
```
Comment (Rugvip) [inline plugins/catalog-backend/src/catalog/DatabaseCatalog.ts:38]: Might as well `YAML.parseAllDocuments` right? Extra work to be done for each doc is [here](https://github.com/eemeli/yaml/blob/3a9327fb224801547f656bdbec20e5f288759aed/src/index.js#L51)

File: plugins/catalog-backend/src/catalog/DatabaseCatalog.ts [hunk 3]
```diff
@@ -15,29 +15,79 @@
...
+  ): Promise<DatabaseCatalog> {
     await database.migrate.latest({
       directory: path.resolve(__dirname, '..', 'migrations'),
       loadExtensions: ['.js'],
     });
 
-    return new DatabaseCatalog(database);
+    const databaseCatalog = new DatabaseCatalog(database, logger);
+
+    const startRefresh = async () => {
```
Comment (Rugvip) [inline plugins/catalog-backend/src/catalog/DatabaseCatalog.ts:69]: I prefer these things outside the constructor, in a start/stop, or open/close pair of methods. Makes it easier to test and lets you check asynchronous pre-work.

File: plugins/catalog-backend/src/catalog/types.ts [hunk 4]
```diff
@@ -17,9 +17,15 @@
 import * as yup from 'yup';
 
 export type Component = {
-  id: string;
+  name: string;
 };
 
+export const componentShape: yup.Schema<Component> = yup
```
Comment (Rugvip) [inline plugins/catalog-backend/src/catalog/types.ts:23]: Don't really think this is where we will end up defining the schema of components 😅. Thinking we pull in a JSONSchema blob eventually, but good to have some validation in place.

**Review by Rugvip — APPROVED** [review]

**Comment by freben** [comment]
Addressing comments in mob as future PR

**Pull request merged**